cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,213 @@
+# CFP-30235: xDS Adapter for ClusterIP Routing
+
+**SIG:** Agent
+
+**Begin Design Discussion:** 2023-01-03
+
+**Cilium Release:** 1.16
+
+**Authors:** Rob Scott <robertjscott@google.com>
+
+## Summary
+
+Add a new xDS adapter in Cilium that could take advantage of some of the
+strengths of xDS, particularly the feedback loop via LRS and the overall
+potential for scalability improvements when adjustments to routing
+configurations don't need to round trip through the Kubernetes API Server.
+This adapter would be an alternative source of endpoints and would not replace
+the existing default behavior of reading directly from Kubernetes APIs.
+
+## Motivation
+
+There are some capabilities that the Kubernetes API Server is not well equipped
+to handle. This includes topology aware routing (or any form of routing that
+requires a feedback loop) and routing to endpoints outside of the cluster.
+
+## Goals
+
+* Introduce a new xDS adapter to Cilium covering ClusterIP routing capabilities.
+* Develop shared interface between Cilium and Kubernetes, KVStore, and xDS as
+  data sources.
+* Provide room for future growth in xDS support beyond ClusterIP routing.
+* Allow the use of any combination of independent data sources (Kubernetes,
+  KVStore, xDS) with Cilium.
+
+## Non-Goals
+
+* Build anything additional on top of this xDS adapter. Although this could be a
+  foundational technology for several new features, those are out of scope for
+  this proposal.
+
+## Background
+
+Several years ago, Matt Klein published a blog post charting a vision for xDS as
+a [universal data plane
+API](https://blog.envoyproxy.io/the-universal-data-plane-api-d15cec7a). Although
+these APIs are primarily used by Envoy and gRPC today, there has been a broader
+goal in the community to enable more widespread usage as part of a truly
+universal data plane API. xDS is CNCF-governed and is gradually transitioning
+content from the
+[envoyproxy/data-plane-api](https://github.com/envoyproxy/data-plane-api) repo
+to [cncf/xds](https://github.com/cncf/xds) repo as part of that overall vision.
+
+In parallel to these efforts by the xDS community, GKE is planning to introduce
+xDS as an additional data source for DPv2 configuration. This feels sufficiently
+generic and helpful that it could be something that could be contributed to
+upstream Cilium. This could be particularly useful for at least two common use
+cases:
+
+1. Supporting Services and Endpoints from outside of the local cluster.
+1. Supporting advanced routing techniques, such as topology aware routing.
```

Comment:
Seems from the LRS stuff that something like:
```
1. Using load information obtained via bidirectional xDS to steer traffic
```
might be useful here? I don't think that comes under "topology", since topology is inherent to the structure of the network, but the load is contingent on external factors.

Reply:
Good point, added. It was always a hope that this kind of load reporting would enable other forms of traffic steering here, added this a use case so it's clear.